tests/integration/src/test/java/org/mule/context/notification/FlowStackTestCase.java

```diff
@@ -0,0 +1,672 @@
+        assertThat(flowStackElements, hasSize(2));
+        assertThat(flowStackElements.get(0), isFlowStackElement("flowInAsync", "/flowInAsync/processors/0 @"));
+        assertThat(flowStackElements.get(1), isFlowStackElement("flowStaticWithAsync", "/flowStaticWithAsync/processors/0/0 @"));
+
+        assertThat(FlowStackAsserter.stackToAssert, areExecutedProcessors(
+                "/flowStaticWithAsync/processors/0 @",
+                "/flowStaticWithAsync/processors/0/0 @",
+                "/flowInAsync/processors/0 @"));
+    }
+
+    @Test
+    public void subFlowStaticWithAsync() throws Exception
+    {
+        FlowStackAsyncAsserter.stackLatch = new CountDownLatch(1);
+        LocalMuleClient client = muleContext.getClient();
+        client.send("vm://in-subFlowStaticWithAsync", new DefaultMuleMessage("payload", muleContext));
+
+        FlowStackAsyncAsserter.stackLatch.await(1, TimeUnit.SECONDS);
+
+        assertThat(FlowStackAsserter.stackToAssert, not(nullValue()));
+        List<FlowStackElement> flowStackElements = FlowStackAsserter.stackToAssert.getElements();
+        assertThat(flowStackElements, hasSize(2));
+        assertThat(flowStackElements.get(0), isFlowStackElement("subFlowInAsync", "/subFlowStaticWithAsync/processors/0/0/subFlowInAsync/subprocessors/0 @"));
+        assertThat(flowStackElements.get(1), isFlowStackElement("subFlowStaticWithAsync", "/subFlowStaticWithAsync/processors/0/0 @"));
+
+        assertThat(FlowStackAsserter.stackToAssert, areExecutedProcessors(
+                "/subFlowStaticWithAsync/processors/0 @",
+                "/subFlowStaticWithAsync/processors/0/0 @",
+                "/subFlowStaticWithAsync/processors/0/0/subFlowInAsync/subprocessors/0 @"));
+    }
+
+    @Test
+    public void flowDynamicWithAsync() throws Exception
+    {
+        FlowStackAsyncAsserter.stackLatch = new CountDownLatch(1);
+        LocalMuleClient client = muleContext.getClient();
+        client.send("vm://in-flowDynamicWithAsync", new DefaultMuleMessage("payload", muleContext));
+
+        FlowStackAsyncAsserter.stackLatch.await(1, TimeUnit.SECONDS);
+
+        assertThat(FlowStackAsserter.stackToAssert, not(nullValue()));
+        List<FlowStackElement> flowStackElements = FlowStackAsserter.stackToAssert.getElements();
+        assertThat(flowStackElements, hasSize(2));
+        assertThat(flowStackElements.get(0), isFlowStackElement("flowInAsync", "/flowInAsync/processors/0 @"));
+        assertThat(flowStackElements.get(1), isFlowStackElement("flowDynamicWithAsync", "/flowDynamicWithAsync/processors/0/0 @"));
+
+        assertThat(FlowStackAsserter.stackToAssert, areExecutedProcessors(
+                "/flowDynamicWithAsync/processors/0 @",
+                "/flowDynamicWithAsync/processors/0/0 @",
+                "/flowInAsync/processors/0 @"));
+    }
+
+    @Test
+    public void subFlowDynamicWithAsync() throws Exception
+    {
+        FlowStackAsyncAsserter.stackLatch = new CountDownLatch(1);
```

Comment:
I would move this initialization to #before and it won't cause any harm on the sync tests